src/imitation/density_baselines.py

```diff
@@ -0,0 +1,267 @@
+      ]
+      self._density_models = [
+          self._fit_single_density(scaled_flat_trans)
+          for scaled_flat_trans in traj_groups_scaled
+      ]
+
+  def _fit_single_density(self, flat_transitions):
+    # This bandwidth was chosen to make sense with standardised inputs that
+    # have unit variance in each component. There might be a better way to
+    # choose it automatically.
+    density_model = KernelDensity(kernel=self.kernel,
+                                  bandwidth=self.kernel_bandwidth)
+    density_model.fit(flat_transitions)
+    return density_model
+
+  def _flatten_trajectories(self, trajectories):
+    flat_trajectories = []
+    for traj in trajectories:
+      obs_vec = traj['obs']
+      act_vec = traj['act']
+      assert len(obs_vec) == len(act_vec) + 1
+      flat_traj = []
+      for step_num in range(len(traj['act'])):
+        flat_trans = self._flatten_transition(obs_vec[step_num],
+                                              act_vec[step_num],
+                                              obs_vec[step_num + 1])
+        flat_traj.append(flat_trans)
+      flat_trajectories.append(flat_traj)
+    return flat_trajectories
+
+  def _flatten_transition(self, obs, act, next_obs):
+    if self.density_type == STATE_DENSITY:
+      return flatten(self.obs_space, obs)
+    elif self.density_type == STATE_ACTION_DENSITY:
+      return np.concatenate([
+          flatten(self.obs_space, obs),
+          flatten(self.act_space, act),
+      ])
+    elif self.density_type == STATE_STATE_DENSITY:
+      return np.concatenate([
+          flatten(self.obs_space, obs),
+          flatten(self.obs_space, next_obs),
+      ])
+    else:
+      raise ValueError(f"Unknown density type {self.density_type}")
+
+  def __call__(self, obs_b, act_b, next_obs_b, *, steps=None):
+    r"""Compute reward from given (s,a,s') transition batch. This handles
+    *batches* of observations, since it's designed to work with VecEnvs.
+
+    Args:
+      obs_b (np.ndarray): current batch of observations.
+      act_b (np.ndarray): batch of actions that agent took in response to those
+        observations.
+      next_obs_b (np.ndarray): batch of observations encountered after the
+        agent took those actions.
+      steps (np.ndarray): number of time steps elapsed in each environment.
+
+    Returns:
+      rew_array (np.ndarray): array of scalar rewards of the form `r_t(s,a,s')
+        = \log \hat p_t(s,a,s')` (one for each environment), where `\log \hat
+        p` is the underlying density model (and may be independent of s', a, or
+        t, depending on options passed to constructor)."""
+    rew_list = []
+    assert len(obs_b) == len(act_b) and len(obs_b) == len(next_obs_b)
+    for idx, (obs, act, next_obs) in enumerate(zip(obs_b, act_b, next_obs_b)):
+      flat_trans = self._flatten_transition(obs, act, next_obs)
+      scaled_padded_trans = self._scaler.transform(flat_trans[None])
```

Comment:
nit: `np.newaxis` instead of None here

Reply:
Does it actually matter?

Reply:
(changed it anyway, I'm just curious why you prefer `np.newaxis`)

Reply:
It doesn't matter, seemed clearer to me. Probably because I keep on forgetting what None does.